Add dfg as a pre-commit hook.

vyasr opened the pull request:
This change allows local and remote runs to handle calls to dfg identically, and removes the need for a separate CI check.

csadorf commented on .github/workflows/pr.yaml:
```diff
@@ -27,6 +27,8 @@ jobs:
     secrets: inherit
     uses: rapidsai/shared-action-workflows/.github/workflows/checks.yaml@branch-23.04
+    with:
+      enable_check_generated_files: false
```
I just tried to figure out what this flag does and from how I understand [this block](https://github.com/rapidsai/shared-action-workflows/blob/43490fb0808ace582f0f09c94e1e54be1e8b7e35/.github/workflows/checks.yaml#L68) it is fairly specific to the files generated by the rapids-dependency-file-checker. Can we maybe rename this flag so that this is a bit more obvious? From the name I would have guessed that it would skip all checks on all modifications applied by "checks" which I don't think would be desirable.

vyasr replied:
I agree, it's not the clearest. However, I would like [to remove the dfg check entirely as a separate reusable action](https://github.com/rapidsai/shared-action-workflows/issues/46), so I don't think it's worth the effort of making further changes here since the flag only really needs to persist until every repo has moved to using the pre-commit hook.

csadorf replied:
Sure.

csadorf commented on .pre-commit-config.yaml:
```diff
@@ -53,6 +53,10 @@ repos:
             pass_filenames: false
             language: python
+    - repo: https://github.com/rapidsai/dependency-file-generator
+      rev: v1.2.0
```
What version scheme does the repo follow? Assuming it is semver, should we adopt the general scheme for GitHub actions repos, where a major tag, e.g. "v1" is moved along the tip of the particular release line?

vyasr replied:
Could you say a bit more? Are you recommending that in addition to the per-release tags, you do something like `git tag -f v1.0` after every minor release to delete what the tag used to point to and repoint to the latest 1.x release? Could you refer to me to where this practice is recommended? I am not necessarily opposed, I'm just not familiar with this approach.

csadorf replied:
Ever since GAs was launched it was officially recommended to use semantic versioning and to have a "major" tag that moves with the tip of the release line using forced tag updates. See point 3 in the [documentation on action versioning](https://github.com/actions/toolkit/blob/master/docs/action-versioning.md#recommendations).

However, this practice of course runs counter to git best practices and is indeed [controversial](https://github.com/actions/toolkit/issues/214). I think this just highlights how using git tags for release management and software lifecycle management is simply not ideal, but for the time being this seems to remain standard practice and from my impression the majority of widely used GitHub actions providers have adopted it.

vyasr replied:
While that may be the recommended best practice for GHA, in this case we're talking about a pre-commit hook. I don't think I've ever seen a hook be configured in this way before, it just so happens that in this case we're using a hook to replace a GHA (that was running the same code under the hood). Does that change your opinion in any way? If we were to change this, would you want to recommend similar changes for all other pre-commit hooks? (That seems like a bigger discussion than we would want to have in this PR, just trying to gauge scope for now)

csadorf replied:
No, I would not recommend that and I think I just got confused reviewing both the action workflow and the pre-commit config at the same time. 😕

vyasr replied:
No worries, thanks for confirming. I think this is ready to go then.

vyasr commented:
Converting to draft pending some discussions on https://github.com/rapidsai/shared-action-workflows/issues/46.

vyasr commented on .pre-commit-config.yaml:
```diff
@@ -53,6 +53,10 @@ repos:
             pass_filenames: false
             language: python
+    - repo: https://github.com/rapidsai/dependency-file-generator
+      rev: v1.2.0
```
```suggestion
      rev: v1.4.0
```

vyasr commented on .pre-commit-config.yaml:
```diff
@@ -53,6 +53,10 @@ repos:
+    - repo: https://github.com/rapidsai/dependency-file-generator
+      rev: v1.2.0
+      hooks:
+          - id: rapids-dependency-file-generator
```
```suggestion
          - id: rapids-dependency-file-generator
            args: ["--clean"]
```

vyasr commented:
Cleaning is now implemented and I have updated the hook version accordingly.

Since there is pending discussion here on the proper versioning scheme for the hook I've converted this PR from draft to be the first one to go through the full review/approval. Once we've decided how to move forward I'll update the corresponding PRs on other RAPIDS repos.

vyasr commented:
/merge
